fw/http.c

```diff
@@ -3349,155 +3369,237 @@ tfw_h1_set_loc_hdrs(TfwHttpMsg *hm, bool is_resp, bool from_cache)
-						     &crlf, NULL);
+			r = tfw_http_msg_expand_data(it, skb_head, &h_mdf,
+						     NULL);
+			if (unlikely(r))
+				return r;
+			r = tfw_http_msg_expand_data(it, skb_head, &crlf,
+						      NULL);
 		} else {
-			r = tfw_http_msg_hdr_xfrm_str(hm, &h_mdf, d->hid,
-						      d->append);
+			r = tfw_http_msg_expand_from_pool(hm, &h_mdf);
+			if (unlikely(r))
+				return r;
+			r = tfw_http_msg_expand_from_pool(hm, &crlf);
 		}
+	}
 
-		if (r) {
-			T_ERR("can't update location-specific header in msg %p\n",
-			      hm);
+	return r;
+}
+
+/**
+ * Add Host header to request.
+ *
+ * RFC 9112 section 3.2.2
+ *
+ * When a proxy receives a request with an absolute-form of
+ * request-target, the proxy MUST ignore the received Host header field
+ * if any) and instead replace it with the host information of the
+ * request-target. A proxy that forwards such a request MUST generate
+ * a new Host field value based on the received request-target rather
+ * than forward the received Host field value.
+ */
+static int
+tfw_http_add_hdr_host(TfwHttpReq *req)
+{
+	int r;
+	TfwHttpMsg *hm = (TfwHttpMsg *)req;
+	static const DEFINE_TFW_STR(host_n, "Host: ");
+	static const DEFINE_TFW_STR(crlf, S_CRLF);
+	TfwStr *host = &req->h_tbl->tbl[TFW_HTTP_HDR_HOST];
+
+	if (test_bit(TFW_HTTP_B_ABSOLUTE_URI, req->flags)) {
+		r = tfw_http_msg_expand_from_pool(hm, &host_n);
+		if (unlikely(r))
 			return r;
-		}
+		r = tfw_http_msg_expand_from_pool(hm, &req->host);
+		if (unlikely(r))
+			return r;
+	} else {
+		r = tfw_http_msg_expand_from_pool(hm, host);
+		if (unlikely(r))
+			return r;
+	}
+
+	return tfw_http_msg_expand_from_pool(hm, &crlf);
+}
+
+/**
+ * Adjust the request before proxying it to real server.
+ */
+static int
+tfw_h1_adjust_req(TfwHttpReq *req)
+{
+	int r;
+	TfwHttpMsg *hm = (TfwHttpMsg *)req;
+	TfwStr *pos, *end;
+	const TfwStr *meth;
+	static const DEFINE_TFW_STR(meth_get, "GET");
+	static const DEFINE_TFW_STR(slash, "/");
+	static const DEFINE_TFW_STR(sp, " ");
+	static const DEFINE_TFW_STR(crlf, S_CRLF);
+	static const DEFINE_TFW_STR(ver, " " S_VERSION11 S_CRLF);
+	const TfwHdrMods *h_mods = tfw_vhost_get_hdr_mods(req->location,
+							  req->vhost,
+							  TFW_VHOST_HDRMOD_REQ);
+
+	req->cleanup = tfw_pool_alloc(hm->pool, sizeof(TfwHttpMsgCleanup));
+	if (unlikely(!req->cleanup))
+		return -ENOMEM;
+	req->cleanup->pages_sz = 0;
+	req->cleanup->skb_head = NULL;
 
-		T_DBG2("updated location-specific header in msg %p\n", hm);
+	tfw_msg_transform_setup(&req->iter, req->msg.skb_head);
+	r = tfw_http_msg_cutoff_headers(hm, req->cleanup);
+	if (unlikely(r))
+		goto clean;
+
+	if (test_bit(TFW_HTTP_B_PURGE_GET, req->flags))
+		/* Rewrite PURGE to GET */
+		meth = &meth_get;
+	else
+		meth = &hm->h_tbl->tbl[TFW_HTTP_METHOD];
+
+	r = tfw_http_msg_expand_from_pool(hm, meth);
+	if (unlikely(r))
+		goto clean;
+
+	r = tfw_http_msg_expand_from_pool(hm, &sp);
+	if (unlikely(r))
+		goto clean;
+
+	if (TFW_STR_EMPTY(&req->uri_path))
```

Comment:
It can be empty  for GET https://tempesta-tech.com:80 for example but should we set it to "/" here?

Reply:
Looks like we should because of RFC 9112 Section 3.2.2